core/src/main/scala/com/banno/kafka/consumer/ConsumerOps.scala

```diff
@@ -372,32 +372,71 @@ case class ConsumerOps[F[_], K, V](consumer: ConsumerApi[F, K, V]) {
   )(
       process: ConsumerRecord[K, V] => F[A]
   )(implicit C: Clock[F], S: Concurrent[F]): Stream[F, A] =
-    for {
-      state <- Stream.eval(
+    Stream
+      .eval(
         C.monotonic
           .flatMap(now =>
             Ref.of[F, OffsetCommitState](OffsetCommitState.empty(now))
           )
       )
-      record <- consumer.recordStream(pollTimeout)
-      a <- Stream.eval(
-        process(record)
-          .onError(_ =>
-            // we can still commit offsets that were successfully processed, before this stream halts, consumer is closed, etc
-            // this is still at-least-once processing, but minimizes the amount of reprocessing after restart
-            state.get.flatMap(s => consumer.commitSync(s.nextOffsets))
-          )
-      )
-      s <- Stream.eval(state.updateAndGet(_.update(record)))
-      now <- Stream.eval(C.monotonic)
-      () <- Stream.eval(
-        s.needToCommit(maxRecordCount, now, maxElapsedTime)
-          .traverse_(os =>
-            consumer.commitSync(os) *>
-            state.update(_.reset(now))
-          )
-      )
-    } yield a
+      .flatMap { state =>
+        consumer
+          .recordStream(pollTimeout)
+          .evalMap { record =>
+            for {
+              a <- process(record)
+                .onError(_ =>
+                  // we can still commit offsets that were successfully processed, before this stream halts, consumer is closed, etc
+                  // this is still at-least-once processing, but minimizes the amount of reprocessing after restart
+                  state.get.flatMap(s => consumer.commitSync(s.nextOffsets))
+                )
+              s <- state.updateAndGet(_.update(record))
+              now <- C.monotonic
+              () <- s
+                .needToCommit(maxRecordCount, now, maxElapsedTime)
+                .traverse_(os =>
+                  // TODO this is probably synchronously blocking a thread, and we should use commitAsync instead
+                  consumer.commitSync(os) *>
+                  state.update(_.reset(now))
+                )
+            } yield a
+          }
+          .onFinalizeCase {
+            // on a clean shutdown, commit offsets of successfully processed records
+            case Resource.ExitCase.Succeeded =>
+              state.get.flatMap(s => consumer.commitSync(s.nextOffsets))
+            // don't commit offsets on Errored or Canceled
```

Comment:
We can commit after an error on a processed record, but not in the general case?  What's different about the state?

Reply:
Committing on a failure of any of the above effects is tricky. This definitely deserves more scrutiny to get right. 

We can commit on a failed `process(record)` but only if we don't include that record's offset in the commit.

We (probably?) wouldn't want to commit if the `consumer.commitSync(os)` failed. Unlikely that a state update, or `C.monotonic` would fail, but it's possible?

I haven't really yet thought through what to do on cancelation.

Reply:
https://github.com/Banno/kafka4s/pull/831/commits/afee61575408d10172eab6d7a2900d09cb4816a2 asks the question: can we commit offsets on any errored stream finalization? Maybe we can... need to stare at the code some more.

The next question: can we commit offsets on any canceled stream finalization? If both answers are "yes" then we can simplify to `.onFinalize(commitNextOffsets)`.

Reply:
One main thing to avoid is ever committing an offset for a record that was not processed successfully. That leads to at-most-once processing, which we cannot have.

We only need at-least-once processing, so it's OK to _not_ commit an offset of a successfully processed record. That leads to potentially reprocessing that record in the future. If `process` is idempotent, at worst that just leads to wasted processing.

Since the state is only updated after the record is processed successfully, maybe it is OK to commit offsets on any error from any failed effect.

Reply:
I agree with the concerns, and that if we adhere to the contract that the state represents success, we're "always committable."

Reply:
I think we can actually commit on any stream finalization: success, error, or cancelation. Thanks for raising the question!